File: go/gf_extern_services/gf_aws/gf_aws_ec2.go
/*
GloFlow application and media management/publishing platform
Copyright (C) 2021 Ivan Trajkovic

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

package gf_aws

import (
	"fmt"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/awserr"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/ec2"
	"github.com/aws/aws-sdk-go/service/autoscaling"
	"github.com/gloflow/gloflow/go/gf_core"
	// "github.com/davecgh/go-spew/spew"
)

//-------------------------------------------------------------
// EC2_SCALE_AUTSCALING_GROUP

func EC2scaleAutoscalingGroup(pAutscalingGroupNameStr string,
	pDesiredCapacityInt int,
	pRuntimeSys         *gf_core.RuntimeSys) *gf_core.GFerror {
	
	sess, err := session.NewSessionWithOptions(session.Options{
		Config:  aws.Config{Region: aws.String("us-east-1")},
		Profile: "default",
	})
	if err != nil {
		gfErr := gf_core.ErrorCreate("failed to create an AWS session object",
			"aws_session_create",
			map[string]interface{}{},
			err, "gf_aws", pRuntimeSys)
		return gfErr
	}

	svc := autoscaling.New(sess)

	//-----------------------
	// UPDATE
	_, err = svc.UpdateAutoScalingGroup(&autoscaling.UpdateAutoScalingGroupInput{
		AutoScalingGroupName: aws.String(pAutscalingGroupNameStr),
		DesiredCapacity:      aws.Int64(int64(pDesiredCapacityInt)),
	})
	if err != nil {
		gfErr := gf_core.ErrorCreate("failed to update an EC2 Autoscaling Group capacity",
			"aws_ec2_autoscaling_scale_error",
			map[string]interface{}{
				"autoscaling_group_name_str": pAutscalingGroupNameStr,
				"desired_capacity_int":       pDesiredCapacityInt,
			},
			err, "gf_aws", pRuntimeSys)
		return gfErr
	}

	//-----------------------
	return nil
}

//-------------------------------------------------------------

func EC2getInfoOnAutoscalingGroup(pAutscalingGroupNameStr string,
	pRuntimeSys *gf_core.RuntimeSys) (int, *gf_core.GFerror) {
	
	sess, err := session.NewSessionWithOptions(session.Options{
		Config:  aws.Config{Region: aws.String("us-east-1")},
		Profile: "default",
	})
	if err != nil {
		gfErr := gf_core.ErrorCreate("failed to create an AWS session object",
			"aws_session_create",
			map[string]interface{}{},
			err, "gf_aws", pRuntimeSys)
		return 0, gfErr
	}

	svc := autoscaling.New(sess)

	//-----------------------
	// GET_INFO
	result, err := svc.DescribeAutoScalingGroups(&autoscaling.DescribeAutoScalingGroupsInput{
		AutoScalingGroupNames: []*string{aws.String(pAutscalingGroupNameStr)},
	})
	if err != nil {
		gfErr := gf_core.ErrorCreate("failed to update an EC2 Autoscaling Group capacity",
			"aws_ec2_autoscaling_describe_error",
			map[string]interface{}{
				"autoscaling_group_name_str": pAutscalingGroupNameStr,
			},
			err, "gf_aws", pRuntimeSys)
		return 0, gfErr
	}

	instancesNumInt := len(result.AutoScalingGroups[0].Instances)

	//-----------------------
	return instancesNumInt, nil
}

//-------------------------------------------------------------

func EC2describeInstancesByTags(p_tags_lst []map[string]string,
	pRuntimeSys *gf_core.RuntimeSys) ([]*ec2.Instance, *gf_core.GFerror) {


	svc := ec2.New(session.New())



	filters_lst := []*ec2.Filter{}
	for _, t_map := range p_tags_lst {
		
		for k, v := range t_map {
			filters_lst = append(filters_lst, &ec2.Filter{
				Name: aws.String(fmt.Sprintf("tag:%s", k)),
				Values: []*string{
					aws.String(v),
				},
			})
		}
	}
	
	input := &ec2.DescribeInstancesInput{
		Filters: filters_lst,
	}


	result, err := svc.DescribeInstances(input)
	if err != nil {
		if aerr, ok := err.(awserr.Error); ok {
			switch aerr.Code() {
			default:
				fmt.Println(aerr.Error())
			}
		} else {
			fmt.Println(err.Error())
		}

		gf_err := gf_core.ErrorCreate("failed to describe ec2 instances with specified tags",
			"aws_ec2_instances_describe_error",
			map[string]interface{}{"tags_lst": p_tags_lst,},
			err, "gf_aws", pRuntimeSys)
		return nil, gf_err
	}




	instances_lst := []*ec2.Instance{}
	for _, r := range result.Reservations {


		instances_lst = append(instances_lst, r.Instances...)
	}


	return instances_lst, nil
}

File: py/src/gf_extern_services/gf_aws/gf_aws_ec2.py
# GloFlow application and media management/publishing platform
# Copyright (C) 2021 Ivan Trajkovic
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import boto3
from botocore.exceptions import ClientError

#-------------------------------------------------------------
def describe_instances_by_tags(p_tags_map,
	p_region_str="us-east-1"):
	
	client = boto3.client("ec2", region_name=p_region_str)

	custom_filter = []
	for k, v in p_tags_map.items():
		custom_filter.append({
			"Name":   f"tag:{k}", 
			"Values": [v]
		})
	
	response = client.describe_instances(Filters=custom_filter)

	aws_instances_lst = []
	for i in response["Reservations"][0]["Instances"]:
		aws_instances_lst.append(i)

	return aws_instances_lst

#-------------------------------------------------------------
# SECURITY_GROUPS
#-------------------------------------------------------------
def get_secgr_id_by_name(p_name_str, p_region_str="us-east-1"):
	assert(isinstance(p_name_str, str))

	client = boto3.client("ec2", region_name=p_region_str)
	
	response = client.describe_security_groups(
		Filters=[{"Name": "group-name", "Values": [p_name_str]}]
	)

	if response["SecurityGroups"]:
		id_str = response["SecurityGroups"][0]["GroupId"]
		return id_str
	else:
		print(f"Security group '{p_name_str}' not found.")
		return None
	
#---------------------------------------------------------------------------------
def add_secgr_ingress_ip(p_ip_cidr_str,
	p_port_int,
	p_sec_gr_id_str,
	p_region_str="us-east-1"):
	assert(isinstance(p_ip_cidr_str, str))
	assert(isinstance(p_port_int, int))
	assert(isinstance(p_sec_gr_id_str, str))

	try:
		client = boto3.client("ec2", region_name=p_region_str)
		
		response = client.authorize_security_group_ingress(
			GroupId=p_sec_gr_id_str,
			IpPermissions=[
				{
					'IpProtocol': 'tcp',
					'FromPort': p_port_int,
					'ToPort': p_port_int,
					'IpRanges': [{'CidrIp': p_ip_cidr_str}]
				}
			]
		)
		
		if response['Return'] is not True:
			raise Exception(f"Ingress rule addition failed: {response}")

	except ClientError as e:
		raise Exception(f"Failed to add ingress rule: {e}")
	
#---------------------------------------------------------------------------------
def delete_secgr_ingress_ip(p_ip_cidr_str,
	p_port_int,
	p_sec_gr_id_str,
	p_region_str="us-east-1"):
	assert(isinstance(p_ip_cidr_str, str))
	assert(isinstance(p_port_int, int))
	assert(isinstance(p_sec_gr_id_str, str))

	try:
		client = boto3.client("ec2", region_name=p_region_str)

		response = client.revoke_security_group_ingress(
			GroupId=p_sec_gr_id_str,
			IpPermissions=[
				{
					'IpProtocol': 'tcp',
					'FromPort': p_port_int,
					'ToPort': p_port_int,
					'IpRanges': [{'CidrIp': p_ip_cidr_str}]
				}
			]
		)
		
		if response['Return'] is not True:
			raise Exception(f"Ingress rule revoking failed: {response}")
		
	except ClientError as e:
		raise Exception(f"Failed to revoke ingress rule: {e}")

#---------------------------------------------------------------------------------
def check_secgr_ingress_ip(p_ip_cidr_str,
	p_port_int,
	p_sec_gr_id_str,
	p_region_str="us-east-1"):
	assert(isinstance(p_ip_cidr_str, str))
	assert(isinstance(p_port_int, int))
	assert(isinstance(p_sec_gr_id_str, str))
	
	try:
		client   = boto3.client("ec2", region_name=p_region_str)
		response = client.describe_security_groups(GroupIds=[p_sec_gr_id_str])

		for permission in response['SecurityGroups'][0]['IpPermissions']:
			if (permission['FromPort'] == p_port_int and
				permission['ToPort'] == p_port_int and
				permission['IpProtocol'] == 'tcp'):
				for ip_range in permission['IpRanges']:
					if ip_range['CidrIp'] == p_ip_cidr_str:
						return True
		return False

	except ClientError as e:
		raise Exception(f"Failed to check ingress rule: {e}")

#---------------------------------------------------------------------------------
# check if the IP/port is allowed in the security group, and if its not add it

def ensure_secgr_ingress_ip(p_ip_cidr_str,
	p_port_int,
	p_sec_gr_id_str,
	p_region_str="us-east-1"):
	assert(isinstance(p_ip_cidr_str, str))
	assert(isinstance(p_port_int, int))
	assert(isinstance(p_sec_gr_id_str, str))

	if not check_secgr_ingress_ip(p_ip_cidr_str, p_port_int, p_sec_gr_id_str, p_region_str):
		add_secgr_ingress_ip(p_ip_cidr_str, p_port_int, p_sec_gr_id_str, p_region_str)